### PR TITLE
feat(#112): drop support for multiple copies of every document

### PR DESCRIPTION
**Pull request by dianabarsan**
# Description

Edits to docs produce updates in dbt models. 
Adds basic DBT models locally and use these for testing. 
Adds two options for deletes - TBD. 

medic/cht-sync#112

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by dianabarsan** [comment]
@witash I've incorporated some of our previously discussed changes: 

1. I'm no longer prefixing the doc ids when inserting into databases, I'm instead spreading the docs between databases. 
2. I've added some local dbt models and made a slight update to the dbt reader to load them (I'm using tarball over http, seemed easiest). 

I'm still not doing deletes, but thought It'd be good to get your thoughts about the above before I proceed. 

Thanks!

**Review by lorerod — CHANGES_REQUESTED** [review]
Thank you, @dianabarsan, for this. I added some minor suggestions.

**Comment by dianabarsan** [comment]
It looks like the username/password is wrong. That's what the 401 status code means.

**Comment by dianabarsan** [comment]
I've added two options for deletes, one that actually deletes data through a post hook and one that just adds the `_deleted` flag. 
I think the latter is easier to cascade and doesn't require the extra delete operation, but it does mean that every dashboard will need to filter out `_deleted` rows. 
I'm not entirely sure how to cascade the full delete, it would require a union with the main table and that seems pretty unclean. 
I'm going to add a "cascade" example for both.

**Comment by lorerod** [comment]
> It looks like the username/password is wrong. That's what the 401 status code means.

I tried with same credentials `curl --location 'https://****:****@brac-clone-for-phil.dev.medicmobile.org:443/medic'` and it responded 200 ok 
I'm doubting myself now.

**Comment by dianabarsan** [comment]
Ok, I'll give this a try.

**Comment by witash** [comment]
> > It looks like the username/password is wrong. That's what the 401 status code means.
> 
> I tried with same credentials `curl --location 'https://****:****@brac-clone-for-phil.dev.medicmobile.org:443/medic'` and it responded 200 ok I'm doubting myself now.

I had the same issue but forgot to mention it.
its something to do with session authentication; commenting out [this](https://github.com/medic/cht-sync/blob/745196cf7bda725d58a80fdeb75fe4ea17b2ddae/couch2pg/src/db.js#L26) line to force it to revert to http basic authentication works (and I guess is also why it works with curl)

**Comment by dianabarsan** [comment]
Wow, thanks for flagging this. 
This is because the session plugin doesn't work with CouchDb 2. I've created an issue to fix it, but in the meantime, would you mind testing this with a CouchDb 3 instance?

**Comment by dianabarsan** [comment]
I've bumped the version of the adapter. @lorerod could you please try again?

**Comment by dianabarsan** [comment]
Hi @njogz ! I've got this PR in the baking that does a bunch of things. The thing that is still unsuccessful is the handling of deletes - which my latest attempt to cascade to next models failed. i would appreciate if you could check it out and add your opinions - and offer solutions if you have a good one for the deletes. 

Thanks!!

**Comment by andrablaj** [comment]
@njuguna-n tagging you as Diana referenced your older GitHub account in the comment above.

**Comment by njuguna-n** [comment]
Thanks @dianabarsan, I will look into the deletes tomorrow.

**Comment by lorerod** [comment]
> I've bumped the version of the adapter. @lorerod could you please try again?

This is working now using the `brac-clone-for-phil.dev.medicmobile.org` CouchDB instance and the .env mentioned [here](https://github.com/medic/cht-sync/pull/115#issuecomment-2181037063)

In the `couch2pg` container log, I can see the majority of the transactions to be successful:
```
2024-06-27 14:42:45 Node.js v20.15.0
2024-06-27 14:34:50 https://admin:9498b55011604aba84f97fcc276114da@brac-clone-for-phil.dev.medicmobile.org:443/medic
2024-06-27 14:34:50 Downloading CouchDB changes feed from 0
2024-06-27 14:35:29 There are 1000 changes to process
2024-06-27 14:35:29 There are 808 deletions and 192 new / changed documents
2024-06-27 14:35:30 Pulled 192 results from couchdb
```
In between, I saw it restarting a couple of times, and at that moment, the logs showed:

```
> 2024-06-27 14:43:43 https://****:****@brac-clone-for-phil.dev.medicmobile.org:443/medic
> 2024-06-27 14:43:43 Downloading CouchDB changes feed from 11219-g1AAAAJ7eJzLYWBg4MhgTmEQTM4vTc5ISXIwNDLXMwBCwxygFFMiQ5L8____szKYkxgYOO1ygWLsZsmmaZaGSdj04DEpSQFIJtnDDWM_AjbM1DjNMinJglTDHECGxcMNY3sANizJLDXR0jKNVMMSQIbVww1jtQYbZpxkaGxhRKo381iAJEMDkAKaNx9iIMsesIEGhmYp5mnGZBm4AGLgfoiBzHcgEWFgaJpKpgsPQAy8DzWwEuLCFMtEA_Nksgx8ADEQGoZMSyGxa2RgbGZkik1rFgCJdqOF
> 2024-06-27 14:43:17   code: 'ENOTFOUND'
> 2024-06-27 14:43:17 }
> 2024-06-27 14:43:17 
> 2024-06-27 14:43:17 Node.js v20.15.0
> 2024-06-27 14:43:43 
> 2024-06-27 14:43:43 /node_modules/node-fetch/lib/index.js:1491
> 2024-06-27 14:43:43                     reject(new FetchError(`request to ${request.url} failed, reason: ${err.message}`, 'system', err));
> 2024-06-27 14:43:43                            ^
> 2024-06-27 14:43:43 FetchError: request to https://brac-clone-for-phil.dev.medicmobile.org/_session failed, reason: getaddrinfo ENOTFOUND brac-clone-for-phil.dev.medicmobile.org
> 2024-06-27 14:43:43     at ClientRequest.<anonymous> (/node_modules/node-fetch/lib/index.js:1491:11)
> 2024-06-27 14:43:43     at ClientRequest.emit (node:events:519:28)
> 2024-06-27 14:43:43     at TLSSocket.socketErrorListener (node:_http_client:500:9)
> 2024-06-27 14:43:43     at TLSSocket.emit (node:events:519:28)
> 2024-06-27 14:43:43     at emitErrorNT (node: internal/streams/destroy:169:8)
> 2024-06-27 14:43:43     at emitErrorCloseNT (node:internal/streams/destroy:128:3)
> 2024-06-27 14:43:43     at process.processTicksAndRejections (node:internal/process/task_queues:82:21) {
> 2024-06-27 14:43:43   type: 'system',
> 2024-06-27 14:43:43   errno: 'ENOTFOUND',
> 2024-06-27 14:43:43   code: 'ENOTFOUND'
> 2024-06-27 14:43:43 }
```
But it restarted successfully, and I didn't see the error anymore. I left it running for 4 hours, and at that point, the `v1.medic` table had 311000 rows.

dbt logs:
```
2024-06-27 18:21:10 22:21:10  Running with dbt=1.8.3
2024-06-27 18:21:11 22:21:11  Registered adapter: postgres=1.8.2
2024-06-27 18:21:12 22:21:12  Found 3 models, 3 data tests, 529 macros
2024-06-27 18:21:12 22:21:12  The selection criterion 'config.materialized:view' does not match any enabled nodes
2024-06-27 18:21:12 22:21:12  
2024-06-27 18:21:12 22:21:12  Concurrency: 1 threads (target='default')
2024-06-27 18:21:12 22:21:12  
2024-06-27 18:21:12 22:21:12  1 of 3 START sql incremental model dbt.contacts ................................ [RUN]
2024-06-27 18:21:12 22:21:12  1 of 3 OK created sql incremental model dbt.contacts ........................... [INSERT 0 341 in 0.50s]
2024-06-27 18:21:12 22:21:12  2 of 3 START sql incremental model dbt.reports ................................. [RUN]
2024-06-27 18:21:13 22:21:13  2 of 3 OK created sql incremental model dbt.reports ............................ [INSERT 0 306 in 0.26s]
2024-06-27 18:21:13 22:21:13  3 of 3 START sql incremental model dbt.persons ................................. [RUN]
2024-06-27 18:21:13 22:21:13  3 of 3 OK created sql incremental model dbt.persons ............................ [INSERT 0 289 in 0.31s]
2024-06-27 18:21:13 22:21:13  
2024-06-27 18:21:13 22:21:13  Finished running 3 incremental models in 0 hours 0 minutes and 1.35 seconds (1.35s).
2024-06-27 18:21:13 22:21:13  
2024-06-27 18:21:13 22:21:13  Completed successfully
2024-06-27 18:21:13 22:21:13  
2024-06-27 18:21:13 22:21:13  Done. PASS=3 WARN=0 ERROR=0 SKIP=0 TOTAL=3
```

**Comment by lorerod** [comment]
@njuguna-n 
I tried this branch with a local **couchdb version 3** instance, and I'm getting these errors in Couch2pg logs:
```
2024-06-28 13:57:00 Node.js v20.15.0
2024-06-28 13:56:57 http://medic:password@localhost:5984/medic
2024-06-28 13:56:57 Downloading CouchDB changes feed from 0
2024-06-28 13:56:58 http://medic:password@localhost:5984/medic
2024-06-28 13:56:58 Downloading CouchDB changes feed from 0
2024-06-28 13:57:00 http://medic:password@localhost:5984/medic
2024-06-28 13:57:00 Downloading CouchDB changes feed from 0
2024-06-28 13:57:04 http://medic:password@localhost:5984/medic
2024-06-28 13:57:04 Downloading CouchDB changes feed from 0
2024-06-28 13:57:04 
2024-06-28 13:57:04 /node_modules/node-fetch/lib/index.js:1491
2024-06-28 13:57:04                     reject(new FetchError(`request to ${request.url} failed, reason: ${err.message}`, 'system', err));
2024-06-28 13:57:04                            ^
2024-06-28 13:57:04 FetchError: request to http://localhost:5984/_session failed, reason: 
2024-06-28 13:57:04     at ClientRequest.<anonymous> (/node_modules/node-fetch/lib/index.js:1491:11)
2024-06-28 13:57:04     at ClientRequest.emit (node:events:519:28)
2024-06-28 13:57:04     at Socket.socketErrorListener (node:_http_client:500:9)
2024-06-28 13:57:04     at Socket.emit (node:events:519:28)
2024-06-28 13:57:04     at emitErrorNT (node:internal/streams/destroy:169:8)
2024-06-28 13:57:04     at emitErrorCloseNT (node:internal/streams/destroy:128:3)
2024-06-28 13:57:04     at process.processTicksAndRejections (node:internal/process/task_queues:82:21) {
2024-06-28 13:57:04   type: 'system',
2024-06-28 13:57:04   errno: 'ECONNREFUSED',
2024-06-28 13:57:04   code: 'ECONNREFUSED'
2024-06-28 13:57:04 }
2024-06-28 13:57:04 
2024-06-28 13:57:04 Node.js v20.15.0
```

To test this, I used CHT Docker Helper for 4.x installed version 4.4.0, which uses "COUCHDB_VERSION=3.3.2" and changed the yml file to expose port 5984. Maybe I'm getting something wrong with my local network. 
If there is an easiest way to try this with couchdb version 3 please let me know.

**Comment by andrablaj** [comment]
@lorerod have you tried running the [CHT > 4.3 in development mode](https://docs.communityhealthtoolkit.org/contribute/code/core/dev-environment/)? The dev env setup worked successfully for me.

**Comment by njuguna-n** [comment]
@lorerod I suspect the issue might be that you have CouchDB running outside the Docker network that Couch2PG is in. Please try using `host.docker.internal` as the host instead of localhost ([docs](https://docs.docker.com/desktop/networking/#use-cases-and-workarounds)).

**Comment by lorerod** [comment]
> @lorerod I suspect the issue might be that you have CouchDB running outside the Docker network that Couch2PG is in. Please try using `host.docker.internal` as the host instead of localhost ([docs](https://docs.docker.com/desktop/networking/#use-cases-and-workarounds)).

Thank you, @njuguna-n. Now it is working! With this, I can confirm it works with Couchdb version 3.3.2

File: tests/e2e-test.spec.js [hunk 1]
```diff
@@ -43,17 +43,88 @@ describe('Main workflow Test Suite', () => {
...
 
-  it('should have data in postgres person table', async () => {
-    const personTableResult = await client.query(`SELECT * FROM ${DBT_POSTGRES_SCHEMA}.person`);
-    expect(personTableResult.rows.length).to.equal(persons().length * dbNames.length);
+  it('should have data in postgres contacts table', async () => {
+    const contactsTableResult = await client.query(`SELECT * FROM ${pgSchema}.contacts`);
+    expect(contactsTableResult.rows.length).to.equal(contacts().length);
   });
 
   it('should have data in postgres data_record table', async () => {
```
Comment (lorerod) [inline tests/e2e-test.spec.js:55]:
```suggestion
  it('should have data in postgres reports table', async () => {
```

File: tests/e2e-test.spec.js [hunk 2]
```diff
@@ -43,17 +43,88 @@ describe('Main workflow Test Suite', () => {
...
-    const dataRecordTableResult = await client.query(`SELECT * FROM ${DBT_POSTGRES_SCHEMA}.data_record`);
-    expect(dataRecordTableResult.rows.length).to.equal(dataRecords().length * dbNames.length);
+    const reportsTableResult = await client.query(`SELECT * FROM ${pgSchema}.reports`);
+    expect(reportsTableResult.rows.length).to.equal(reports().length);
+  });
+
```
Comment (lorerod) [inline tests/e2e-test.spec.js:59]: Can we add a validation also for `persons` table? To test all the models defined for testing in`/tests/dbt/models`?

File: tests/e2e-test.spec.js [hunk 3]
```diff
@@ -43,17 +43,88 @@ describe('Main workflow Test Suite', () => {
...
+  it('should process document edits', async () => {
+    const report = reports()[0];
+    const contact = contacts()[0];
+    await editDoc({ ...report, edited: 1 });
+    await editDoc({ ...contact, edited: 1 });
+
```
Comment (lorerod) [inline tests/e2e-test.spec.js:65]: Also, please add a specific `persons` edition and the corresponding validations here.

File: tests/e2e-test.spec.js [hunk 4]
```diff
@@ -43,17 +43,88 @@ describe('Main workflow Test Suite', () => {
...
+    const preDelete = await client.query(`SELECT * FROM ${pgSchema}.persons where _id = $1`, [person._id]);
+    expect(preDelete.rows.length).to.equal(1);
+
+    await deleteDoc(person);
+    await delay(6); // wait for CHT-Sync
+    await delay(6); // wait for DBT
+    await delay(6); // wait for DBT
+
+    const modelContactResult = await client.query(`SELECT * FROM ${pgSchema}.contacts where _id = $1`, [person._id]);
+    expect(modelContactResult.rows[0]._deleted).to.equal(true);
```
Comment (lorerod) [inline tests/e2e-test.spec.js:113]:
```suggestion
    expect(modelContactResult.rows.length).to.equal(1);
    expect(modelContactResult.rows[0]._deleted).to.equal(true);
```

File: tests/e2e-test.spec.js [hunk 5]
```diff
@@ -43,17 +43,88 @@ describe('Main workflow Test Suite', () => {
...
+
+    const postDelete = await client.query(`SELECT * FROM ${pgSchema}.persons where _id = $1`, [person._id]);
+    expect(postDelete.rows[0]._deleted).to.equal(true);
```
Comment (lorerod) [inline tests/e2e-test.spec.js:116]:
```suggestion
    expect(postDelete.rows.length).to.equal(1);
    expect(postDelete.rows[0]._deleted).to.equal(true);
```

**Review by lorerod — CHANGES_REQUESTED** [review]

File: tests/e2e-test.spec.js (hunk 3 above)
Comment (lorerod) [reply tests/e2e-test.spec.js:65]: Also, every person is a contact, right? So, the edit should be propagated to the person, too.

Comment (njuguna-n) [reply tests/e2e-test.spec.js:73]: Yes, they should be propagated. I have added an assertion for this.

File: tests/e2e-test.spec.js (hunk 4 above)
Comment (njuguna-n) [reply tests/e2e-test.spec.js:124]: This is handled in this [PR](https://github.com/medic/cht-sync/pull/118)

File: tests/e2e-test.spec.js (hunk 5 above)
Comment (njuguna-n) [reply tests/e2e-test.spec.js:127]: This is also addressed in this [PR](https://github.com/medic/cht-sync/pull/118)

**Review by lorerod — APPROVED** [review]
Thanks @njuguna-n

**Pull request merged**